GH-46734: [C++] Arrow Flight SQL ODBC layer

alinaliBQ opened the pull request:
### Rationale for this change
* Build Arrow Flight SQL ODBC DLL from `flightsql` and `odbcabstraction` component. 
* The ODBC driver will be available in Windows. Future support for macOS/Linux is work-in-progress.

### What changes are included in this PR?
* WIP - Arrow Flight SQL ODBC DYLIB for macOS.
* This PR acts as a placeholder for our ODBC code changes. Our team will cherry-pick from this PR and raise smaller PRs for the community to review and merge. (See [comment](https://github.com/apache/arrow/pull/46099#issuecomment-3367554851)) 
* Community is welcomed to leave comments on this PR as well, though comments on the smaller PRs will be addressed faster

### Are these changes tested?
Yes, on local Windows and macOS environment. 

### Are there any user-facing changes?
N/A

This PR addresses the following issues:
* GitHub Issue: #46734

jduo commented on cpp/src/arrow/flight/sql/odbc/entry_points.cc:
```diff
@@ -0,0 +1,43 @@
+#include "arrow/flight/sql/odbc/odbc_api.h"
+#include "arrow/flight/sql/odbc/visibility.h"
+
+SQLRETURN SQL_API SQLAllocHandle(SQLSMALLINT type, SQLHANDLE parent,
```
Do you need to use an extern "C" block for ODBC API functions?

jduo commented on cpp/src/arrow/flight/sql/odbc/odbc_api.cc:
```diff
@@ -0,0 +1,56 @@
+    */
+    return SQL_INVALID_HANDLE;
+  }
+}  // namespace arrow
```
There should be a newline at the end of each file.

alinaliBQ commented:
@aiguofer I have opened a draft PR for the ODBC layer.

alinaliBQ commented:
@zeroshade , @assignUser , @ianmcook  Follow-up on the May-07 Arrow meeting, this is the draft PR that Rob and I have for creating the initial connection in the ODBC layer. It is based on top of @jbonofre 's [PR#40939](https://github.com/apache/arrow/pull/40939). Some changes in this PR are planned to be moved to [PR#40939](https://github.com/apache/arrow/pull/40939), such as the `vcpkg.json` update.

alinaliBQ commented on cpp/vcpkg.json:
This is what `vcpkg.json` change will look like. Please help review this change

jbonofre commented:
@alinaliBQ why not just using the original PR/branch ? Just to not loose changes/comments.

alinaliBQ commented:
> @alinaliBQ why not just using the original PR/branch ? Just to not loose changes/comments.

Yup you're right, for `flightsql-odbc` contribution, folks should use the apache#40939 PR, and this PR has higher priority for review than apache#46099. After apache#40939 PR is merged, we can go back to look at this apache#46099 PR.

alinaliBQ commented:
Hello @lidavidm , this is the draft PR that is based on top of @jbonofre 's https://github.com/apache/arrow/pull/40939. It's still a draft PR but just wanted to give you a heads up

alinaliBQ commented:
Hi @lidavidm, please let me know if you prefer the Arrow Flight SQL ODBC contribution PRs like this one to be split into smaller PRs. It would be nice to know in advance so my team can prepare for the split, thank you!

kou commented:
Yes. I prefer smaller PRs to larger PRs because smaller PRs are easier to review than larger PRs.

(BTW, I'm not sure whether this PR is a smaller PR or not...)

lidavidm commented:
Yes, smaller PRs would be preferable. (To be frank, I have yet to actually review the first PR in full, but I'd rather get the paperwork done and get it merged ASAP, then we can go over it in detail afterwards.)

alinaliBQ commented:
@kou @lidavidm Ok sounds good! I will rebase `Bit-Quill:apache-odbc` on top of `apache:main` branch after https://github.com/apache/arrow/pull/40939 is merged, so the size of this PR will be smaller compared to what it shows right now. Then I plan to open new smaller PRs to split this PR.

alinaliBQ commented:
Hi David @lidavidm and Kou @kou, there's been a change regarding my last comment, we (Rob @rscales & I) will still work on splitting the PR into smaller ones and look into the `flightsql-odbc` comments from David on https://github.com/apache/arrow/pull/40939 (and any other comments for Arrow ODBC Flight SQL), but that will be after we finish developing the ODBC layer to have the full functionality. 
This PR#40939 will stay in draft state until we finish developing the ODBC driver, but any changes in this PR are ready for community to review. 

TL;DR, we would like to finish the ODBC layer development first, and we will begin addressing the community comments after that.

Please let me know if you have any questions. :)

alinaliBQ commented on dev/release/rat_exclude_files.txt:
```diff
@@ -13,6 +13,8 @@ ci/vcpkg/*.patch
 CHANGELOG.md
 cpp/CHANGELOG_PARQUET.md
 cpp/src/arrow/c/dlpack_abi.h
+cpp/src/arrow/flight/sql/odbc/odbc.def
+cpp/src/arrow/flight/sql/odbc/install/install_amd64.cmd
```
The Apache Ignite [install_amd64.cmd](https://github.com/apache/ignite/blob/927ead13979937195aca6ffb717942fb477dc1b6/modules/platforms/cpp/odbc/install/install_amd64.cmd) file does not have a license on it, so I assume `.cmd` files don't need licenses. Same logic for `.def` files; the Apache Ignite [module.def](https://github.com/apache/ignite/blob/38a4804/modules/platforms/cpp/odbc/module.def) file does not have a license.

kou replied:
Could you add license header to all files expect that don't support comment.
We can use `rem ...` in `.cmd` and `; ...` in `.dev` for comments.

lidavidm commented:
OK - sounds good, please ping me when that is ready for review

kou commented:
Could you fix a build error https://github.com/apache/arrow/issues/46576 before the ODBC layer development?

lidavidm commented on cpp/src/arrow/flight/sql/odbc/CMakeLists.txt:
```diff
@@ -15,7 +15,149 @@
+
 add_custom_target(arrow_flight_sql_odbc)
 
+# Ensure fmt is loaded as header only
+add_compile_definitions(FMT_HEADER_ONLY)
```
C++20 has std::format - is that insufficient?

lidavidm commented on cpp/src/arrow/flight/sql/odbc/README:
Please make sure all files have the copyright header.

lidavidm commented on cpp/src/arrow/flight/sql/odbc/flight_sql/address_info.h:
```diff
@@ -19,8 +19,10 @@
+#include "arrow/flight/sql/odbc/odbcabstraction/include/odbcabstraction/platform.h"
+
 #include <sys/types.h>
+#include <cstdint>
```
Can we keep headers sorted? standard library headers should go together (and go above project header includes)

lidavidm commented on testing:
Was this intentional?

alinaliBQ replied:
nope, let me fix this

lidavidm commented on cpp/src/arrow/flight/sql/odbc/flight_sql/accessors/timestamp_array_accessor.cc:
```diff
@@ -85,6 +88,10 @@ RowStatus TimestampArrayFlightSqlAccessor<TARGET_TYPE, UNIT>::MoveSingleCell_imp
     int64_t& value_offset, bool update_value_offset,
     odbcabstraction::Diagnostics& diagnostics) {
+  // Times less than the minimum integer number of seconds that can be represented
+  // for each time unit will not convert correctly.  This is mostly interesting for
```
Can we detect & error?

lidavidm commented on cpp/src/arrow/flight/sql/odbc/odbcabstraction/include/odbcabstraction/odbc_impl/encoding_utils.h:
```diff
@@ -77,4 +78,39 @@ inline size_t ConvertToSqlWChar(const std::string& str, SQLWCHAR* buffer,
+  return std::string(utf8_str.begin(), utf8_str.end());
+}
+
+inline std::string SqlStringToString(const unsigned char* sqlStr,
+                                     int32_t sqlStrLen = SQL_NTS) {
```
```suggestion
inline std::string SqlStringToString(const unsigned char* sql_str,
                                     int32_t sql_str_len = SQL_NTS) {
```
etc.

lidavidm commented on cpp/src/arrow/flight/sql/odbc/install/install_amd64.cmd:
What about this is specific to amd64?

alinaliBQ replied:
The install script specifically installs the 64-bit Windows version (there is no 32-bit driver)

alinaliBQ commented:
Hi @lidavidm @kou  we are close to start working on community code review comments to prepare for merging the PR. Since this PR is quite big, here is my proposal:

* create new PRs that extract smaller commits from this PR to merge to `main`, then rebase this PR #46099 onto main after smaller new PR is merged, and repeat the process until this PR #46099 is small enough to be merged to `main` directly

Does this approach look good to you? 

Feel free to tag other maintainers who may be interested in reviewing the code.

kou commented:
Yes. In general, I prefer smaller PRs. :-)

lidavidm commented:
Yes, this is nearly impossible to review as is :sweat_smile:

alinaliBQ commented:
https://github.com/apache/arrow/pull/47517 has been created

alinaliBQ commented:
Posted PR:
https://github.com/apache/arrow/pull/47645

kou commented:
Yes. Let's use the approach.

alinaliBQ commented:
Sounds good. When the draft PRs are ready for a review, the team will tag folks as before

alinaliBQ commented:
I hope this makes it easier for reviewing: If folks see any general comments on our draft PRs, feel free to mention that it is a `general` comment, and our team will work on these on all applicable PRs. Please don't feel the need to post the general comments on all the applicable PRs. 
For in-progress PRs that don't have all known general comments addressed, we will label them with `WIP`.
Thanks for helping to review/merge our PRs!

jmao-denver commented:
@alinaliBQ  Do you know anything about this: https://docs.influxdata.com/influxdb3/enterprise/visualize-data/powerbi/#install-the-arrow-flight-sql-odbc-driver?

alinaliBQ commented:
Hi @jmao-denver, the ODBC driver published on Influx Data is built from the branch `apache-odbc` in this PR, albeit it’s probably not on the latest commit since we update the PR weekly. Since the code is Apache-licensed, my understanding is that companies are allowed to host and publish the MSI installer on their sites. 

Arrow Flight SQL ODBC hasn't been officially released by the Arrow community yet, and my team is working with the Arrow maintainers to get the Windows ODBC driver merged.

alinaliBQ committed:
Add Arrow Flight SQL ODBC driver

Co-authored-by: alinalibq <alina.li@improving.com>
Co-authored-by: justing-bq <justin.gossett@improving.com>
Co-authored-by: vic-tsang <victor.tsang@improving.com>

ODBC commits squashed into 1 to make rebasing to main easier.

 Add initial framework for odbc dll

- Add ARROW_FLIGHT_SQL_ODBC option. If we set `ARROW_FLIGHT_SQL_ODBC=ON`, the flightsql odbc folder will be built
- Add odbc api layer for entry_point.cc
- builds odbc dll file, with ODBC APIs exported in odbc.def

Address James' comments

Fix `odbcabstraction` build errors and partially fix `flightsql_odbc` errors

Fix boost-variant not found error
- Adding dependencies from odbc/vcpkg.json to cpp/vcpkg.json

- Fix whereami.cc and .h dependency; ported lates code

Update whereami.cc

- use `long` instead of `int64`. Fixed namespace issues.

- PR CI fix: Add `parquet-testing` back

Partial build fix for `flight_sql` folder

- Replaced `namespace arrow` and `namespace odbcabstraction` with `using namespace ...`

- fix flight_sql_connection.cc

Fix `util::nullopt` to use `std::nullopt`

- fix std::optional

- fix BufferReader

- Fix GetSchema

- fix json_converter.cc

- partial fix configuration.h

-  partial fix get_info_cache.cc

- Fix winsock build error

- Comment out `flight_sql` files that cannot build
   - Comment out configuration and unit tests
   - Comment out get info cache and system trust store

Create initial odbc tests folder

Implement SQLAllocEnv

Fix cmake build

Implement SQLFreeEnv

Fix rest of build errors from `flightsql-odbc`

- Fix get info errors

- Fix for configuration window

  - added odbcinst library

- Fix system trust store

- unit test fixes

   - Add dependency of ARROW_COMPUTE. `arrow/compute/api.h` is used in
     `flight_sql`. Adding `ARROW_COMPUTE=ON` during build fixed run time unit
     tests failures.

Implement SQLAllocConnect and SQLFreeConnect

Fix build issue from static flight sql driver

Lint and code style fixes

Re-add deleted submodule parquet-testing

clang-format lint fix

cpplint lint fix

Exclude whereami in rat exclude list

C++/CLI lint fix

Update parquet-testing to match commit from `main`

Address Kou's comments

ODBC directory lint fixes

Catching the lint fixes outside of `flightsql-odbc` code

Fix build warnings that get treated as error

Implement SQLSetEnvAttr and SQLGetEnvAttr

Implement use of ExecuteWithDiagnostics

Doxygen Error Fixes and Address comments from Kou and James

Address comments from Kou

- Updates License.txt
- Update cmake toolchain
- Move whereami to `vendored`
- Use string_view instead of NOLINT std::string

Remove `whereami.cc` from arrow util build

We are building whereami.cc as part of odbc

Fix include headers to replace <> with ""

Address comments from James

Implement SQLGetDiagField

SQLDriverConnect, SQLConnect and SQLDisconnect

Implement stubs for SQLGetInfo, SQLGetDiagField and SQLGetDiagRec

Separate RegisterDsn and UnregisterDsn from windows build

Update code to save driver value from connection string

Add ReadMes for ODBC and tests

Fix test issues with string_view

Address code reviews

Update entry_points.cc to fix build issue

Remove Dremio references

Use emplace properly

Address comment from Rob and add SQLDisconnect test case

Add odbc.def and cmd file to rat_exclude

Nit - remove duplicate lines

Accidentally committed the change during git rebase

Nit - remove usage of nullptr

DSN window implementation

Add licenses to `.cmd` and `.def` files

Implement SQLGetDiagRec

Build ODBC in Windows Workflow

Tests are skipped for now.

Updates for SQLGetDiagFieldW

Enable Driver Logging

Add todo to update logging system later

Add logs

Implement ODBC API with debug logging

Enable mock test

* Add todo for noauth validation

* mock server with token auth

Add tests

* run same test with both modes

* Enable ODBC tests in workflow

* Switch current test cases to use FlightSQLODBCTestBase

So the tests can be skipped when `TEST_CONNECT_STR` is not set.

* Change tests to run on both mock and remote modes

Wrap usage of TEST_CONNECT_STR where possible

* Rename test fixtures and make connection string functions virtual

* Fix lint issue

* Attempt to enable ODBC build on Windows platforms

* Attempt to fix clang64 and MinGW errors

* Attempt to register ODBC

* Address James' comments

Use constant string for token

* use ServerMiddleware to validate token

Fix connection issues to DBT Labs

PopulateCallOptions before making a connection

Fix dsn window bug with advance properties

Fix seg fault issue from empty string

Implement SQLAllocStmt

Follow-up DBT Labs connection fix

Implement SQLGetDiag Rec and Field for statement

Unicode support for DSN ODBC APIs

* Let compiler append `W` to ODBC APIs where applicable.

Initialize Kernel functions

As per changes from #46261, we need to initialize Kernel library explicitly to get the functions registered

Implement SQLSetConnectAttr and SQLGetConnectAttr

ODBC Test Segfault Fix

* Move `connection_test.cc` to last in `CMakeLists.txt`, this resolves the seg fault. The seg fault was not reproducible on my local environment, even when I use the msys2 shell.
* Rename test executable to flight_sql_odbc_test

SQLExecDirect implementation

* SQLGetStmtAttr stub implementation

* Fix missing break statement in SQLGetDiagRec

* Run ShutdownProtobufLibrary as part of test environment

* Add comment for GH-46889

* Pass call options to executed prepared statement

`call_options_` contains the authentication token which is needed

Implement SQLGetInfo

SQLGetInfo Workflow Fixes

SQLFetch & SQLGetData Implementation

SQLGetStmtAttr stub implementation

Stub call for SQLGetStmtAttr

Enable statement handle allocation

Implement SQLExecDirect - test hang issue

Update odbc_api.cc

Run `ShutdownProtobufLibrary` after all ODBC tests

Resolves test hanging problem

Fix missing break statement in SQLGetDiagRec

Add tests

Lint fixes

Run ShutdownProtobufLibrary as part of test environment

Add debug messages

Draft code

Remove drafts

Add comment

Add comment for GH-46889

Update statement test headers

Pass call options to executed prepared statement

`call_options_` contains the authentication token which is needed

SQLFetch & SQLGetData initial impl

EVERYTHING before this commit is for SQLExecDirect

SQLFetch and SQLGetData TestSQLExecDirectSimpleQuery test

Make GetData() return SQLRETURN

`TestSQLExecDirectDataQuery` for remote and mock servers

Remove unneeded test case

Add more data types

* Add SQL_GUID in getCTypeForSQLType

* Add data types in query

* Add checks for columns 1 to 25

* Update statement_test.cc

Add varbinary test

* Implement SQLNumResultCols

Will leave tests for later

* Implement SQLRowCount

* Implement SQLMoreResults

* Work on date retrieval fix

The `TestSQLExecDirectDataQuery` checks for date is not working.

* Fix date retrieval error by importing fix from Paul

* imported fix from https://github.com/dremio/flightsql-odbc/commit/d44d862bd07f05328f5fcfd68a8f40357aa072fa, an Apache-Licensed repository

* Add more checks for get data

* Add more tests

* Continue work on float truncation test

* Disable float truncation test

* Address comments from James

- do not return errors for invalid rowCountPtr and columnCountPtr
- add static cast for columnCount
- Add tests with SQL_C_DEFAULT as target type
- Add test checks for indicator
- Fix SQL_NUMERIC, SQL_BIT, SQL_FLOAT target type

* Fix build issues in CI

* More fix for CI

* Add test for invalid buffer length

SQLPrepare & SQLExecute Implementation

* Add tests
* Address comments from Rob

Code Style fixes

* Move ifdefs outside of test cases

* use `stmt`

* Move logs to first line

* fix test for `SQL_DRIVER_HSTMT`

* Use camel case for tests

Implement SQLGetStmtAttr and SQLSetStmtAttr

SQLBindCol Implementation

* SQL_UNBIND implementation + tests

* Add tests for `SQL_ATTR_ROW_ARRAY_SIZE`

* Add check for `SQL_ATTR_ROWS_FETCHED_PTR`

Enable TestCloseConnectionWithOpenStatement test case

SQLFetchScroll Implementation

* Add checks for SQL_ATTR_ROWS_FETCHED_PTR

* SQLFetchScroll is supposed to return the number of rows fetched using SQL_ATTR_ROWS_FETCHED_PTR

SQLExtendedFetch Implementation

* Implement rowCountPtr and rowStatusArray for SQLExtendedFetch

* Add tests for SQLExtendedFetch

* `SQLExtendedFetch` doesn't return `SQL_SUCCESS_WITH_INFO` for error state 22002.

* Add tests for `SQL_ROWSET_SIZE`

* Address comments from James

GH-46584 Iterate endpoint

* use suggestion from James for one-liner change to return `Status::OK` directly

* fix for iterating endpoints, it is based on JDBC's fix for the same bug

* save value of `FlightClientOptions`

* Use `TestFlightServer`

Add `arrow_flight_sql_shared` to enable usages for `TestFlightServer`

* Fix build errors from missing `gmock`

* Add checks for array data

Update flight_sql_stream_chunk_buffer_test.cc

Add `FlightStreamChunkBufferTest` unit test

Draft test with `FlightSQLODBCMockEndpointTestBase`

* Reference GH-47117 and Clean up code

* Add driver and DSN to built-in properties to not pass driver/dsn as attributes to server

* Allow `=` in values inside connection strings, fixes connectivity problems

* Address comments from Rob

SQLColumns Implementation

* Initial impl for SQLColumns

* Add test for SQLColumns

* Add columns test with all supported column types

- mock server doesn't support schema

* Address comments from James

- Test different data type return value for SQLColumns
- Add todo comment for SQLDescribeCol tests

---------

Co-authored-by: rscales <robscales@icloud.com>

Fix bug of DSN not read properly

* Fix reading DSN bug

Use a simpler, more robust way to load the DSN

SQLColAttribute implementation

* SQLColAttribute initial impl

* Switch to use `GetAttributeSQLWCHAR` to be unicode-compatible

* Add SQLColAttribute tests and fix bugs in `flightsql-odbc` impl

* Fixed bug with incorrect column count. It was returning column count + 1.
* Fixed bug with incorrect numPrecRadix. It was returning SQL_NO_TOTAL (maps to -4) for non-numeric columns, but the correct value is 0.
* Fixed bug with unsigned column to return true for unsigned columns (non-numeric columns or unsigned numeric columns) and false for signed columns (numeric columns)
* Fixed bug with incorrect non-concise data type return value for date time type types. The correct return is SQL_DATETIME for all date time types

* Address comments from James

* use const for `ConvertToWString`
* add nullptr check in odbc_api.cc layer
* Add support for `SQL_COLUMN_LENGTH`, `SQL_COLUMN_PRECISION`, and `SQL_COLUMN_SCALE`. The driver will return the same values for ODBC2 and ODBC3.
* Add tests for ODBC v2 SQLColAttributes values
* I converted tests that can be run on remote servers where possible

* Address comment from Rob

Implement SQLTables

Use C++ 20 standard for ODBC

* Add support for building with C++20

* Upgrade spdlogs version as an attempt to fix build issues

* Fix issue of missing unique_ptr definition

Add tests for SQLMoreResults

Implement SQLNativeSql

Add diagnostic records

* remove `SQLErrors`, as the driver manager is supposed to map `SQLErrors` to `SQLGetDiagRec`

Implement tests for SQLNumResultCols

SQLCloseCursor Implementation

* fix error state to return the correct state 24000.

* add tests for SQLCloseCursor and SQLFreeStmt(SQL_CLOSE)

Add SQLColumns and SQLColAttribute tests with nullptr

* Modify SQLTables test to run on both mock and remote
* reorder error msg alphabetically

Implement tests for SQLRowCount

Fix merge conflicts

Adjust searchable return value

based on GH-46920 fix

Fix `getCTypeForSQLType` return for interval types

- Fix `getCTypeForSQLType` function to return the correct `C type` for interval SQL data types. Previously the function was checking for interval `C type` and returning interval `SQL type`, which was the opposite of the intended functionality.

SQLGetTypeInfo implementation

* Add tests for SQLGetTypeInfo

* fix bug of createParams returning empty string instead of null
* fix bug of non-concise data type being returned for datetime/interval
* fix bug of longvarchar not converted to wlongvarchar when driver has unicode enabled

* Address comments from James and add checks

*  test SQL_ALL_TYPES in ODBC 2.x.
*  test SQL_TYPE_* in ODBC 2.x, the driver manager reports invalid data type error.
*  test SQL_* datetime in ODBC 3.x for backwards compatibility.
*  add check for valid/invalid data type.

Fix segfault issue from empty metadata

* Use empty map in bug fix

* Address comments from James

Implement SQLDescribeCol

Add SQLError Tests

* add test to free null handles. Without handle value initialization, segfault error was seen

Move `SQLGetDiagField` and `SQLGetDiagRec` tests to `errors_test.cc`

* Address comments from James

* Add ODBC Ver 2 tests

* update test name to indicate if error handling is from driver manager.
* add tests for warnings.

* fix lint errors.
* remove SQL_ATTR_APP_ROW_DESC that is not applicable to Environment Attribute.

Add tests for SQLGetFunctions

* Using `SQL_FUNC_EXISTS` macro fixed the issue of `api_exists` not read correctly

* Fix SQLGetTypeInfo naming for ODBC ver 2 tests

* Add more tests

Add unsupported API checks.
Add `TestSQLGetFunctionsODBCVer2`.
Add `TestSQLGetFunctionsCheckSingleAPI`.

* Update reset value

* Add SQLDescribeCol function check

Add additional SQLDescribeCol test cases

Add Descriptor support in SQLAllocHandle and SQLFreeHandle

* Descriptor allocation initial impl

* Add descriptor handle tests

* Add diagnostic error test for descriptor handle

* the error is from driver manager as our driver doesn't have descriptor-specific APIs implemented yet

* Add SQLGetDescField test from driver manager

* Add doxygen doc comment

Fix connection issues servers that require data in handshake

* revert back to passing empty string in handshake

Fix bug of system trust store not loaded properly

* Bug: system trust store has default value of "true" regardless of "encryption" value.
* But if encryption is set to false, system trust store cannot be used, so system trust store should be saved as false if user does not enable encryption.
* It may confuse the user if they see encryption is set to false but system trust store is set to true. In this case, the driver will not do system trust store verification. The DSN window should reflect this accurately.

CPack ODBC Windows msi installer

Add ARROW_FLIGHT_SQL_ODBC_INSTALLER environment variable to enable creation of Windows installer.

* Use CPack + WiX to create a `msi` installer.
* run command `cpack` under the build directory to generate the installer.

Register ODBC Driver on Windows in installer

* In `wxs`, cannot use `package`, need to use `fragment` instead

* Use component as feature will automatically be generated

* Set Patch version + add installer instructions

Use Arrow logo banner

* replaces default banner with Arrow logo banner

Indicate Driver Company and Version on Driver Manager

* use versioninfo.rc.in template

* use `1 VERSIONINFO` for it to work properly

* Set version variables

* Add `.rc` to gitignore

* Add `@` variables to rc template

Return nameLengthPtr value as length in characters

Allow spaces while parsing Table Type

mac changes

added changes to build in MacOS

Enable ODBC build on Windows Glib & Ruby workflows

* Fix `boost::get` issue on GLib workflow

* Fix GLib build issue with conflicts

Compile entry_points.cc before odbc_api.cc due to conflict from sql.h and flight/types.h.

Fix build warnings treated as error on Windows

SQLDescribeCol expects `bufCharLen` to be of type `SQLSMALLINT`, so build warnings occur when `bufCharLen` is set to `SQLINTEGER`.

Add more types for Array Conversion

Set column attribute SQL_DESC_TYPE_NAME

[Refactor] Remove reference of `Configuration` in odbc_connection.cc

Remove reference of `Configuration` in odbc_connection.cc which is in library `odbcabstraction`. The reasoning is that `Configuration` is in `arrow_odbc_spi_impl` and this is a library that `odbcabstraction` depends on. I moved the logic for reading DSN values into `odbc_api.cc` instead, since that is a place that will result in no conflicts. The ODBC driver now does not save the DSN value in the connection string if the DSN value is put after Driver value.

* Address comments from James

Simply logic to check for first key value pair only

Add TODO for non-UTC time zone support

* Currently, the driver only supports UTC time zone

Replace `spdlogs` with Arrow's Internal Logging

* Add logging README

* register kernel function conditionally

   * Resolves errors of kernel function already registered

Logging files are not supported since GLOG is not enabled on Windows in Arrow repo. We can work + test on log file support on macOS/Linux phase

Remove  `RUN_ALL_TESTS` as they are not needed

Previously, tests on Windows were not being run due to https://github.com/apache/arrow/issues/47434. So we added  `RUN_ALL_TESTS`  as a workaround in early May, before https://github.com/apache/arrow/issues/47434 was raised.
Now that https://github.com/apache/arrow/issues/47434 is resolved, we can remove  `RUN_ALL_TESTS` as they are not needed.

Other changes:
* Renamed `arrow_odbc_spi_impl_test` to `odbc_spi_impl_test` so the test executable will be  `arrow_odbc_spi_impl_test.exe` instead of  `arrow_arrow_odbc_spi_impl_test.exe`

Fix Function & Variable Naming

Fix member variable naming

Address code review comments for Arrow#40939

* Reword comment on null check

Addresses comment https://github.com/apache/arrow/pull/40939#discussion_r2099120394

* Use `std::memcpy` instead of `memcpy`

Addresses comment https://github.com/apache/arrow/pull/40939#discussion_r2099120764
* add include cstring

* Remove warpdrive mentions

* Fix lint

* Address Justin's comments

Combine utils and define utils namespace

Use arrow::flight::sql::odbc namespace

Update namespaces driver::flight_sql and driver::odbcabstraction to arrow::flight::sql::odbc.

Fix seg fault from register log

PR to keep apache-odbc in sync with arrow main branch.

I found that, if the driver registers logs before compute kernel registration, segfault somehow occurs at when register kernel function is being used. I am not sure the root cause, but registering the logs after compute kernel registration seems to resolve the issue.

Refactor odbcabstraction and odbc_impl

flight_sql is renamed to odbc_impl. The files inside odbcabstraction have been moved to odbc_impl and odbcabstraction is removed.

Replace boost::variant with std::variant

* Remove `boost-optional` and `boost-variant` dependencies from vcpkg.json

* Replace boost::variant with std::variant

Fix EXPECT_EQ() to have expected on the left and actual on the right

Code style naming convention

Remove underscores in test names

Sync arrow_flight_testing changes from Arrow::main branch

Fix test failure due to typo

Fix `FlightSQLODBCRemoteTestBase.TestSQLExecDirectDataQueryDefaultType` test failure due to typo
The precision should be 38 in the original code. I think it got changed to 0 by accident somewhere.

Disable tests with `DISABLED_`

* Enable `TestSQLGetInfoDriverHdesc`

We missed this test before, enabling the test now

* Disable tests with `DISABLED_`

Fix test assertions

Define Test Fixtures

* Connect/disconnect in SetUp/TearDown
* Define Test Fixtures

Resolve ODBC CI Build Errors

* Fix `ConfigDSNW` not found error

* Fix `TestFreeNullHandles` test

`ConnectionTest` sets up the handles, so the `this->...` handles will not be null at beginning of test.

Define New Base Test Fixtures

Skip tests without connecting/disconnecting

Address general test code review comments

* Use RAII helper for allocating and freeing env/conn handles

Avoids duplicated code

Move test helper functions to anonymous namespace

Add `[[nodiscard]]` for ODBC APIs

Addresses community comment:
https://github.com/apache/arrow/pull/47763/files#r2450014186

Remove `using List=` in test suite definitions

Use static_cast for `SQLWCHAR` in type info test

Use static_cast for `SQLWCHAR` in tables test

* use mutable arrays for places where characters cannot be const

Use static_cast for `SQLWCHAR` in columns test

Update comment

Use static_cast for `SQLWCHAR` in SQLGetInfo test

* Address Justin's comments

* Add fix for skipping the allocation of handles

Enable ODBC build in MacOS CI

ODBC Standalone MSVC Build CI

Remove DataSet

Fix cache key

Use odbc-specific cache key

Fix Lint and Add odbc registration step

Link appropriate GitHub issues that blocks ODBC tests

Disable test phase and add schedule

Run everyday 7 am Vancouver time

Enable ODBC build on MSVC CI

Code Clean up and enable ODBC tests in CI

Still need to modify ODBCUtilEnvironment if we decide to use it. Still need to add ODBC V2 support so a different env and conn is used for ODBC 2 tests.

Draft enable ODBC global setup/teardown

Run ODBC test once outside of test script

If ODBC test is run using MinGW Shell, segfault occurs.
I am not getting any seg fault on my local MSVC Windows when I run the tests without the bash script. But if Windows CI breaks from running the standalone exe then I will look into this

Prepend vcpkg to search vcpkg before `<prefix>/lib/cmake`

Since we are installing dependencies on vcpkg, if `lib/cmake` is searched first, then cmake will look into that directory and use the wrong paths.

Convert VCPKG Windows path to MSYS path

Set `VCPKG_ROOT` in test phase

Fix ODBC dll name

Use `arrow_flight_sql_odbc.dll` instead of `libarrow_flight_sql_odbc.dll` which is the naming convention used on MinGW Windows

Link ODBC library on all platforms

On my local MSVC Windows machine, Visual Studio is used to build without needing to link ` ${ODBCINST}` explicitly, its behavior might be different from Ninja which is what CI uses.
`${ODBCINST}` is likely needed by Linux as well, so adding it for all platforms.

Attempt to resolve `arrow-compute-grouper-benchmark` build issue

I think `if(ARROW_FLIGHT_TEST_LINKAGE STREQUAL "static")` might be more appropriate here, as I am seeing build issues due to both dynamic and static linking occurring. I see in https://github.com/apache/arrow/commit/59903d089ec5b1766e1622d94c1f618b539b205d, this check is used for `arrow-filesystem-s3fs-benchmark`

Disable `UNITY_BUILD` for ODBC test due to conflict on `sqlite_sql_info.cc`

On some workflows, unity build is set to ON to make the build faster. This is an attempt to resolve the build issue.

Remove debug messages

Fix lint

Add `sql_info_undef.h` to sqlite_sql_info.cc

Undefine duplicates in SQLGetInfo

Add `#pragma once` to odbc_test_suite.h

To avoid redefinition error

Attempt to resolve conflict with sql/types.h

Attempt to include Arrow headers before ODBC headers to avoid conflict with arrow/flight/sql/types.h

[GH-48084] Replace boost::optional with std::optional

Addresses comment https://github.com/apache/arrow/pull/40939#discussion_r2099118497
Replace boost::optional with std::optional in ODBC codebase

https://github.com/apache/arrow/issues/48084

Fix lint

Remove `BOOST_SOURCE=BUNDLED` to use vcpkg's dynamic link to boost

Since we need to use link to boost dynamically, we need to remove the bundled boost library flag.

Add debug messages.

Remove `ARROW_BOOST_USE_SHARED = OFF`

Since we have a release build, can enable boost as shared.

With `ARROW_BOOST_USE_SHARED = OFF`, I am getting error
```
D:\a\arrow\arrow\build\cpp\vcpkg_installed\x64-windows\include\boost/filesystem/config.hpp(96): fatal error C1189: #error:  Must not define both BOOST_FILESYSTEM_DYN_LINK and BOOST_FILESYSTEM_STATIC_LINK
```

Also increased time limit, since 2 hours don't seem to be enough to finish the vcpkg build

Change to release build

Getting
```
orc.lib(Exceptions.cc.obj) : error LNK2038: mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in unity_2_cxx.cxx.obj
orc.lib(Exceptions.cc.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MDd_DynamicDebug' in unity_2_cxx.cxx.obj
```
when building with debug and presumably ARROW_ORC

Set ARROW_BOOST_USE_SHARED to OFF

`ARROW_BOOST_USE_SHARED` is restored to `OFF`, which according to Windows doc should help with the debug build now.

Retore value of `ARROW_BUILD_BENCHMARKS`, though noting it is set to `OFF` in GLib MSVC workflow.

Add VCPKG set up

Borrowed from the Glib workflow

Add `/EHsc` flag

* Add back `CMAKE_CXX_STANDARD: "17"`

Remove `CMAKE_CXX_STANDARD` 17

* reason: gtest can't be used with C++17. Arrow project doesn't support C++ 17 yet.

Extend run-time to 2hr

windows-mingw also has timeout of 2hr

Empty commit to trigger CI

Specify `VCPKG_BINARY_SOURCES` and `VCPKG_DEFAULT_TRIPLET`

Specify VCPKG_TARGET_TRIPLET as x64 windows

Set ARROW_DEPENDENCY_SOURCE to VCPKG

To make it easier to manage dependencies

Enable static build on MSVC

`ARROW_BUILD_BENCHMARKS` prevents Flight from being built

Remove `ARROW_BOOST_USE_SHARED`

Having static vs. shared issue

Enable Flight & Flight SQL for ODBC

Enable ODBC build on MSVC CI

Enable regular ctest tests

Remove undef items

Add concurrency and permissions to odbc yml

Create cpp_odbc.yml

Fix architecture in CI

* Add ODBC installer MSVC CI

Mac Setup ODBC ini Script

* added setup script for mac and platform folders

updated read me windows file path

updated windows folder related file paths

added license

Fix Lint issues

Add install ODBC script

Update install_odbc_ini.sh

This script depends on the ODBC install script.

Add TODO note and update DSN

* Fix typo in README

---------

Fix lint

Fix lint

Update flight_sql_stream_chunk_buffer.cc

Update statement_test.cc

Move MacOS ODBC workflows to cpp_odbc.yml

Fix Lint

Fix `ODBC::GetSqlWCharSize()`

Add back missing checks for error handle tests

SQLError implementation in macOS

* SQLError implementation for macOS
Disable `SQLGetConnectOption` on macOS. Still need to check if SQLGetConnectOption is needed on macOS Excel.

Update type_info_test.cc

Update columns_test.cc

Remove unneeded component code

- Removed unneeded code for `COMPONENT`
- Use `install(PROGRAMS ${CMAKE_INSTALL_SYSTEM_RUNTIME_LIBS}` to include MSVC redistributable DLLs instead of picking them up with `unspecified` components.
- Use better way to include ODBC dll files: replace DLL filtering with `install(TARGETS`

Fix deadlock error from C++ 20

* Fix ODBC deadlock issue

`absl::SetMutexDeadlockDetectionMode(absl::OnDeadlockCycle::kIgnore);` fixes ODBC deadlock issue. We turn off the detection as the deadlock originated from upstream projects.

* Fix ODBC Extra slow CI issue

Attempt to resolve caching issue with ODBC CI

Add write permission to store vcpkg cache

Check nuget paths in Ruby Glib and ODBC

Run vcpkg install after vcpkg is set in GLib

Enable vcpkg verbose in GLib

* confirmed `vcpkg install` command returns 403 error

Add ARROW_HOME

Remove `ARROW_SIMD_LEVEL` as it is not required for ODBC.

Remove debug msgs

Cannot get vcpkg in cpp_build to show debug messages

Move location of `packages: write`

Attempt to add `ARROW_DEPENDENCY_USE_SHARED - off`

Revert "Move location of `packages: write`"

This reverts commit b62e04ed6d88296217bd8ed05d67ee2fb0b927c2.

Revert "Attempt to add `ARROW_DEPENDENCY_USE_SHARED - off`"

This reverts commit 5fdf4251df18a4f2c1b7c697f99fe91ef6e1b9aa.

Add GLib MSVC build to C++ Extra

TEMP disable ODBC build

Disable cmake and enable build

Re-enable cmake 4.1.2 and use install_vcpkg.sh to install vcpkg

Clean up test changes

Push after previous run is complete.

Revert "Re-enable cmake 4.1.2 and use install_vcpkg.sh to install vcpkg"

This reverts commit dfea3774de18f764ee6bcba8293d347e4e358813.

Trigger CI

Apply Kou's suggestion for nuget timeout

Remove `C++ ODBC` workflow

* missed item from rebase

Remove changes added from rebase

Co-Authored-By: Victor Tsang <victor.tsang@improving.com>
Co-Authored-By: Alina (Xi) Li <alina.li@improving.com>
Co-Authored-By: vic-tsang <victor.tsang@improving.com>

Build ODBC statically on macOS

* Link ODBC library statically on unix

Change unit tests to re-use code. Testing on windows is needed.

Now seeing error
```
Library not loaded: /usr/local/opt/grpc/lib/libgrpc++.1.76.dylib
```
on Excel.

Attempt to unlink arrow_flight.proto

Now I see ODBC.dylib has 100MB, and error from Excel is
```
Library not loaded: /usr/local/opt/libiodbc/lib/libiodbc.2.dylib
```
instead of looking for flightsql.

`iodbctest` command outside of Excel works.

in-progress link arrow statically for `flightsql-odbc`

* both ODBC driver and `arrow_odbc_spi_impl` need to link Arrow statically for tests to work.
* getting error
`File already exists in database: Flight.proto`
since the ODBC layer is linking dynamically while I successfully got `arrow_odbc_spi_impl` to link to Arrow statically

Draft for resolving `File already exists in database: FlightSql.proto` error

Note: make ODBC available on Excel first, test with `iodbctest`, worry about ODBC test executable later.

Add unix build (same as Windows build)

* Attempted to link static libraries with `arrow_odbc_spi_impl`, failed as it caused linking errors and rendered the ODBC driver unusable.

Undo changes to resolve broken driver

Resolves dependency errors

Attempt to switch to static build

Still getting error
```
[iODBC][Driver Manager]dlopen(/Library/ODBC/arrow-odbc/libarrow_flight_sql_odbc.2300.0.0.dylib, 0x0006): Library not loaded: @rpath/libarrow_flight_sql.2300.dylib
```

In-progress attempt to build ODBC statically on unix systems

In-progress changes to enable static odbc_impl library

pushing the changes just for saving my code. Currently odbc test still has double registration issue, work on this next. I think it should be solvable by either adding library flags or just using odbc spi shared. Because `odbc_spi_impl_test` works. And the ODBC tests worked just fine before I switched to static odbc spi impl.

Fix for gtest missing issue

Attempting to fix for issue:
IMPORTANT NOTICE - DO NOT IGNORE:
This test program did NOT call testing::InitGoogleTest() before calling RUN_ALL_TESTS().
This is INVALID. Soon Google Test will start to enforce the valid usage.
Please fix it ASAP, or IT WILL START TO FAIL.

Issue fixed with adding `${ARROW_TEST_LINK_LIBS}`. It is needed alongside `arrow_flight_testing_shared`.

IN-PROGRESS build static test on macOS

* Trying to have 2 builds. One `arrow_odbc_spi_impl_static` for static build and one `arrow_odbc_spi_impl_shared` for a shared lib that links Arrow dynamically
---
only committing the approach that worked.
Trying to work from `flight_sql_odbc_test` to debug the real issue.

Link 3rd party dependencies statically

Fix 3rd party headers cannot find issue for:
- boost xpressive and beast headers
- rapidjson headers

Resolve unneeded link for ODBC

----
Moving `set(APPEND CMAKE_FIND_LIBRARY_SUFFIXES ".a")` up didn't seem to have any effect

----
Fix for static gRPC connection issue:
export GRPC_DEFAULT_SSL_ROOTS_FILE_PATH=/etc/ssl/cert.pem

Fix cares `cannot modify alias target` issue

Make arrow_odbc_spi_impl static lib

In progress for getting gprc linked statically

Add debug msgs in add_arrow_lib BuildUtils.cmake

Remove commented out code in odbc lib

Fix build issue from odbc impl tests

Added draft code for Link Arrow libs statically on Unix for unit tests

Add commented out code`ARROW_DEPENDENCY_USE_SHARED` in ODBC cpp yml

Uncomment when it is verified to work.

In-progress Fix ODBC test build and make it run.

Build executable from scratch

Revert "Build executable from scratch"

This reverts commit a51dc395d69c50b42090f50c22cdc9247d3baf9f.

Dummy test

dummy test without using ODBC passed without double registration issue

Make ODBC test run

Issue to resolve:
[libprotobuf FATAL /path/to/arrow/cpp/debug-build/_deps/protobuf-src/src/google/protobuf/extension_set.cc:100] Multiple extension registrations for type "google.protobuf.MessageOptions", field number 1000.
unknown file: Failure
C++ exception with description "Multiple extension registrations for type "google.protobuf.MessageOptions", field number 1000." thrown in SetUp().

Fix: only link test executable directly with ODBC shared dylib. Do not link test executable with odbc spi impl dylib. And change ODBC to link odbc spi impl dylib publicly instead of privately

in-progress changes to accomodate macos CI

Test ODBC CI - to be reverted

Can revert later

Only install ODBC dependencies in mac

As dependencies are static and built from source

Uninstall absl as attempt to resolve build on Intel

Add C++ standard 20

Remove absl header from intel

Use static linking in dependency install step

Use bundled boost for static linking

In-progress Fix Windows build

Fix macOS (after Windows build)

Remove unneeded ODBC link

Resolves the issue of ODBC on macOS being dynamically linked to ODBC

Fix ODBC double proto issue on macOS

Need to check if Windows CI passes

Clean up PR

- Remove dummy tests
- Remove unneeded changes

Revert "Test ODBC CI - to be reverted"

This reverts commit 711b3e0863ebbf5fe2c5cb1401fbaf72eed58aa4.

Remove unneeded code

Attempt to revert c-ares change

Fix `set_target_properties can not be used on an ALIAS target.` error.

Example of failed run in forked repo:
https://github.com/Bit-Quill/arrow/actions/runs/21459770235/job/61809740997?pr=153#step:7:540

Fix descriptor pointer tests on static macOS

Fix mimalloc issue

Fixes issue:
```
mimalloc: assertion failed: at "arrow/cpp/debug-build/mimalloc_ep-prefix/src/mimalloc_ep/include/mimalloc/internal.h":658, _mi_ptr_page
  assertion: "p==NULL || mi_is_in_heap_region(p)"
zsh: abort      debug/arrow-flight-sql-odbc-test
```

Enable ODBC tests for static build

Add iodbc link to macOS build on ODBC layer

Add iodbc link to macOS build on ODBC Test

Add iodbc link to macOS build on ODBC layer did not resolve issue of unixodbc being picked in CI

* Fix segfault at SQLError

* work on Justin's comments

Fix rebase conflicts

Remove code added during conflict resolution

Implement system_dsn for Mac

Address community comments for `install_odbc_ini.sh`

Add separate release & debug workflows for MacOS

Disable DSN default values on macOS

* Remove DSN default values on macOS

* Work on Justin's comment

Refactor test setup/teardown to reuse connection across tests

* Define Mock Server SetUp/TearDown at Environment level

* Refactor test setup/teardown to reuse connection across tests

* In-progress fix DSN write issue on macOS without `sudo`

* Getting errors of `SQLSetConfigMode` flagged as `void`,

```
error: cannot initialize return object of type 'bool' with an rvalue of type 'void'
  502 |   ASSERT_TRUE(SQLSetConfigMode(ODBC_BOTH_DSN));
```

* Clean Up PR

* Remove unneeded changes

---------

Fix rebase errors

Change `GetInfo` disabled tests to start new connection

* Enable disabled tests

* Fix getinfo tests to use standalone connection

Windows CI to Support ODBC DLL & MSI Signing

* Add draft code for CI A and CI B

Attempt workflow dispatch

Only ODBC Windows original workflow should run.
Later need to add `github.event_name != 'workflow_dispatch' ||` to all existing workflows after uncomment

Use `GITHUB_REF_NAME` directly via push

Add `workflow_dispatch` definitions

Add `ODBC Windows Upload DLL`

Use common ODBC Windows environment variables

Use ODBC as composite action

Create cpp_odbc.yml

Initial draft

temp disable test step

Temp disable non-ODBC Windows workflows

* Clean Up Code

* Remove comments

* Fix Installer path for MSI

Fix issue with `secrets.GITHUB_TOKEN`

Co-Authored-By: Alina (Xi) Li <alinal@bitquilltech.com>

justing-bq committed:
Get test suite building on Linux

justing-bq committed:
Use SqlWCharArrLen() for wide char array length on Linux

justing-bq committed:
Fix tests on Linux

alinaliBQ committed:
Fix ODBC Linux test segmentation fault

* Resolve segfault issue on Linux and pass `ARROW_TEST_LINKAGE=static`

Pass `ARROW_TEST_LINKAGE=static`

* Clean up

alinaliBQ committed:
Disable Large Memory Tests locally

alinaliBQ committed:
Enable Remote Testing With Dremio Instance

* Disable non-odbc workflows

* Install `netcat` and set `ARROW_FLIGHT_SQL_ODBC_CONN`

* Attempt to enable docker network

Update compose.yaml

Add executable bash script

* add commands for testing Dremio instance value

* Change to use `host.docker.internal`

* Change test to run remote test

Update cpp_extra.yml

* Use `localhost` as hostname

* Use `dremio_container` as host

* Clean up PR

* Test remote instance

* CI can connect to remote instance, revert "Test remote instance"

This reverts commit db975bfd30490e21e21b4667f8a004f49347e0ca.

* Disable ODBC Debian build

* Seeing segfault at ODBC test. Can probably be solved by passing test linkage = static like we did with Ubuntu. Since we don't have capacity to work on Debian now, we can close this.

justing-bq committed:
Reuse connections across test suite

Co-Authored-By: Alina (Xi) Li <96995091+alinaliBQ@users.noreply.github.com>

alinaliBQ committed:
macOS `.PKG` installer for Intel and ARM

* Initial draft for macOS .pkg installer

* in-progress for `install_odbc`

* Remove `$HOME` from registration script

* Generate .pkg installer and attempts to fix installer

* Attempt to fix doc not seen
* Attempt to fix ODBC registration script

* Fix installer script and doc

* Rename `install_odbc_ini.sh` to `postinstall`

* Reuse `install_odbc.sh` script inside `postinstall`

* Fix to generate macOS installer

- Check $(pwd)/build/cpp

* Clean up PR and todos

* Update format to re-use code

* Use `install_odbc_ini.sh` script to install DSN

Keep a lightweight `postinstall` file for macOS

* Update install_odbc_ini.sh execution access

* Address Justin's comment

alinaliBQ committed:
Add debug workflow for ODBC Linux

* Add debug workflow for ODBC Linux

Update cpp_extra.yml

* apply static test linkage to ODBC only

alinaliBQ committed:
ODBC Linux rpm installer support 

* Implement RPM installer

Revert "TEMP - Disable non-ODBC items"

This reverts commit efa8d84c696688432ff43689e84123e22a0e47f7.

fix formatting

Remove todos

ODBC rpm Installer wrap up

* installing `rpm` package fixed the `cpack` command issue
* Fix cpack command on CI

(merge into last commit when ready) Change to use `awk` so script works on both platforms

Comment out `find` and `tree` steps

Change build path to be under /arrow

Since docker is mounted, changing build path to be in `/arrow` will allow host machine to access docker build contents

TEMP - Disable non-ODBC items

Add commands to find build folder

Attempt to build RPM installer

* not sure if `/build/cpp` can be accessed outside of docker. Hopefully it should.
* add `rpm` dependency install

Indicate RPM in workflow run

Add rpm `postinstall` (not tested)

Remove merge conflict code

* Enable ODBC installer build in CI instead of `compose.yaml`

alinaliBQ committed:
Only Run ODBC-related tests on CI

* Only Run ODBC-related tests on CI

* Increase timeout for macOS

* Work on Justin's comment

justing-bq committed:
Initialize Arrow Compute in global test environment

justing-bq committed:
Set ODBC handles as test fixture fields

justing-bq committed:
ODBC connection properties documentation for unix

justing-bq committed:
Update README

alinaliBQ committed:
Linux DEB ODBC installer support

* Initial draft of DEB installer

* reuse script for RPM, and rename script to indicate `postinst` (postinstall)

* Add `file` to dockerfile to enable Debian installation

* Fix component settings with DEB

* Fix package name

* Clean up comments

* Fix file name

* Fix postinst variable for DEB

* Fix upload artifact file name and doc file name
